File: ModularAritmos.py
# ver. Tue/02/Jan/2024
#
# Made by: CyberCoral
# ------------------------------------------------
# Github:
# https://www.github.com/CyberCoral
#

class Complex:
    def __init__(self, a, b):
        self.real = a
        self.imaginary = b
    def real(self):
        return self.real
    
    def imaginary(self):
        return self.imaginary

def real_mod(a, b):
    '''
    Does
    "a mod b"
    for all real a and b.
    '''
    if isinstance(a, float) != True and isinstance(a, int) != True:
        raise TypeError("'a' must be a float or int")
    if isinstance(b, float) != True and isinstance(b, int) != True:
        raise TypeError("'b' must be a float or int")
    if b == 0:
        raise ZeroDivisionError("Division by zero in function.")
    
    mod = (a / b - int(a / b)) * b
    return mod

def one_mod(a):
    '''
    Does
    "a mod 1"
    for all real a.
    '''
    mod = a - int(a)
    return mod

def floor0(a):
    '''
    Does floor(a),
    method 1 with
    a mod 1.
    '''
    return a - one_mod(a)

def floor1(a):
    '''
    Does floor(a)
    method 2 with
    int(a)
    '''
    return int(a)

def complex_mod(a: list, b: list, complex_: bool = True):
    '''
    Does
    "a mod b"
    for complex numbers
    a and b.
    List form (2 elements each).
    '''
    if isinstance(a, list) != True:
        raise TypeError("'a' must be a list.")
    elif len(a) != 2:
        raise ValueError("'a' must have 2 elements.")
    elif [isinstance(i, int) != True and isinstance(i, float) != True for i in a].count(True) != 0:
        raise TypeError("'a's elements must be int or float.")

    if isinstance(b, list) != True:
        raise TypeError("'b' must be a list.")
    elif len(b) != 2:
        raise ValueError("'b' must have 2 elements.")
    elif [isinstance(i, int) != True and isinstance(i, float) != True for i in b].count(True) != 0:
        raise TypeError("'b's elements must be int or float.")

    if isinstance(complex_, bool) != True:
        raise TypeError("complex_ must be a bool.")

    a1, a2, b1, b2 = a[0], a[1], b[0], b[1]
    dem, c_dem = b1 ** 2 + b2 ** 2, b1 ** 2 - b2 ** 2
    result = []

    if dem == 0:
        raise ZeroDivisionError("The denominator must not be zero.")

    f1 = floor1(((a1 * b1) + (a2 * b2)) / dem)
    f2 = floor1(((a1 * b2) - (b1 * a2)) / dem)

    g1 = (a1 * c_dem - (2 * a2 * b1 * b2)) / dem
    g2 = (a2 * c_dem - (2 * a1 * b1 * b2)) / dem

    result = [ g1 - (f1 * b1) + (f2 * b2), g2 + (f1 * b2) + (f2 * b1) ]

    if complex_ != True:
        return result
    return complex(result[0], result[1])

def MOD(num: list, real_nums: list, imag_nums: list, complex_: bool = True):
    '''
    Does
    'num mod a+bi'
    for as many items in
    real_nums and imag_nums are
    (they must have the same number
    of elements), includes complex
    mod
    '''
    if isinstance(num, list) != True:
        raise TypeError("'num' must be a list.")
    elif len(num) != 2:
        raise ValueError("'num' must have 2 elements.")
    elif [isinstance(i, int) != True and isinstance(i, float) != True for i in num].count(True) != 0:
        raise TypeError("'num's elements must be int or float.")

    if isinstance(real_nums, list) != True:
        raise TypeError("'real_nums' must be a list.")
    elif len(real_nums) < 0:
        raise IndexError("The list must not be empty.")
    elif [isinstance(i, int) != True and isinstance(i, float) != True for i in real_nums].count(True) != 0:
        raise TypeError("'real_nums's elements must be int or float.")

    if isinstance(imag_nums, list) != True:
        raise TypeError("'imag_nums' must be a list.")
    elif len(real_nums) < 0:
        raise IndexError("The list must not be empty.")
    elif [isinstance(i, int) != True and isinstance(i, float) != True for i in imag_nums].count(True) != 0:
        raise TypeError("'imag's elements must be int or float.")

    if len(real_nums) != len(imag_nums):
        raise IndexError("The num parts must have the same number of elements.")

    if isinstance(complex_, bool) != True:
        raise TypeError("complex_ must be a bool.")

    for i in range(len(real_nums)):
        num = complex_mod(num, [real_nums[i], imag_nums[i]], False)
    
    if complex_ != True:
        return num
    return complex(num[0], num[1])


    
    

    


    

File: ModularAritmos.c
#include <stdio.h>
#include <stdlib.h>
#include <math.h>

/*

Created by CyberCoral
Version Tue/02/Jan/2024

-----------------
Github:
https://github.com/CyberCoral
-----------------
*/

/* A structure simulating complex numbers */
struct complex_reduced{
    long double real;
    long double imag;
};

/* Modular arithmetic functions */

double real_mod(double a, double b){

    double variable[1];

    double a1 = a;
    double b1 = b;

    if (b1 == 0){
        printf("\nb1 cannot be zero.");
        exit(1);
    };

    double f1 = (a1 / b1 - floor( a1 / b1 )) * b1;
    variable[0] = f1;

    return *variable;
};

double one_mod(double a){

    double one_module[1];

    double a1 = a;

    one_module[0] = a1 - floor(a1);

    return *one_module;
};

int floor0(double a){
    int floor = a - one_mod(a);
    return floor;
};

double* complex_mod(double a[2], double b[2]){

    //double complex_module[2];
    double* complex_module = malloc(16);

    double a1 = a[0];
    double a2 = a[1];
    double b1 = b[0];
    double b2 = b[1];
    double dem = pow(b1,2) + pow(b2,2);
    double c_dem = pow(b1,2) - pow(b2,2);

    if(dem == 0){
        printf("\nThe denominator cannot be zero.");
        exit(1);
    };

    double f1 = floor(((a1 * b1) + (a2 * b2)) / dem);
    double f2 = floor(((a1 * b2) - (b1 * a2)) / dem);
    double g1 = (a1 * c_dem - (2 * a2 * b1 * b2)) / dem;
    double g2 = (a2 * c_dem - (2 * a1 * b1 * b2)) / dem;

    complex_module[0] = g1 - (f1 * b1) + (f2 * b2);
    complex_module[1] = g2 + (f2 * b2) + (f1 * b1);

    return complex_module;

};

double* MOD(double a[2], double* r_arr, double* i_arr){
    double* arr = malloc(16);
    arr[0] = a[0];
    arr[1] = a[1];

    double check[2];
    check[0] = sizeof(r_arr) / sizeof(r_arr[0]);
    check[1] = sizeof(i_arr) / sizeof(i_arr[0]);

    if(check[0] != check[1]){
        printf("\nr_arr and i_arr do not have the same number of elements.\n");
        exit(1);
    };

    double next[2];
    next[0] = r_arr[0];
    next[1] = i_arr[0];

    arr[0] = complex_mod(arr,next)[0];
    arr[1] = complex_mod(arr,next)[1];

    for (int i=1;i<check[0];i++){
        next[0] = r_arr[i];
        next[1] = i_arr[i];

        arr[0] = complex_mod(arr,next)[0];
        arr[1] = complex_mod(arr,next)[1];
    };
    return arr;
};

/*You gotta start out of something :) */


int example(void){

    printf("Some examples of modular arithmetic operations:\n");

    double x = 45;
    double y = 999;

    double a = real_mod(x,y);

    printf("\nreal mod: ");
    printf("%lf",a);

    double z = 5.4321;

    int z0 = floor0(z);

    double b = one_mod(z);

    printf("\none_mod: ");
    printf("%lf",b);

    printf("\nfloor function: ");
    printf("%i",z0);

    double a1[2];
    a1[0] = 1;
    a1[1] = 1;

    double a2[2];
    a2[0] = 2;
    a2[1] = 2;

    double* c;
    c = complex_mod(a1,a2);

    printf("\ncomplex_mod: ");
    printf("%lf%+lfi", c[0],c[1]);

    free(c);

    double r_arr[1] = {1};
    double i_arr[1] = {0};

    double comp[2] = {4,5};

    double* d = MOD(comp, r_arr,i_arr);

    printf("\ncomplex_MOD: ");
    printf("%lf%+lfi", d[0],d[1]);

    free(d);

    printf("\nThe program finished successfully :D");

    return 0;

}

int main(){
    example();
    // Pause till enter is pressed.
    getchar();
};
